openhcl/underhill_core/src/dma_manager.rs

```diff
@@ -0,0 +1,131 @@
+//use std::sync::{Mutex, Weak};
+
+use memory_range::MemoryRange;
+use std::{collections::HashMap, sync::{Arc, Mutex}};
+use page_pool_alloc::PagePool;
+use user_driver::{memory::MemoryBlock, vfio::VfioDmaBuffer};
+use user_driver::lockmem::LockedMemorySpawner;
+
+#[derive(Clone)]
+pub struct GlobalDmaManager {
+    _physical_ranges: Vec<MemoryRange>,
+    _bounce_buffers_manager: Vec<MemoryRange>,
+    //clients: Mutex<Vec<Weak<DmaClient>>>,
+    //client_thresholds: Mutex<Vec<(Weak<DmaClient>, usize)>>,
+
+    page_pool: Option<PagePool>,
+    clients: HashMap<String, Arc<DmaClient>>,
+}
+
+impl GlobalDmaManager {
+    pub fn new(page_pool: Option<PagePool>) -> Self {
+        GlobalDmaManager {
+            _physical_ranges: Vec::new(),
+            _bounce_buffers_manager: Vec::new(),
+            //clients: Mutex::new(Vec::new()),
+            //client_thresholds: Mutex::new(Vec::new()),
+            page_pool: page_pool.clone(),
+            clients: HashMap::new(),
+        }
+    }
+
+    pub fn create_client(&mut self, pci_id: String) -> DmaClient {
```

Comment:
I'm not sure we want to _clone_ DmaClients, as it seems reasonable to me that there is at _most_ one client per pci_id. I'd rather here we make the Arc external, ie don't implement clone, but instead return an `Arc<DmaClient>`